test/webhook-manager.js

```diff
@@ -0,0 +1,146 @@
+   *         .expect(200)
+   *
+   *       processedEventIds[event.id] = true
+   *     }
+   *
+   *     const timestamps = events.map(e => e.created)
+   *     const maxTimestamp = _.max(timestamps) - 1 // to be sure to fetch processed events
+   *     return maxTimestamp ? maxTimestamp - 1 : null
+   *   }
+   *
+   * @param {Boolean}  [params.webhookFetchInterval = 2000] - specify the duration in milliseconds for events fetching
+   *   if isWebhookSimulated is true
+   * @param {Function} [params.createWebhook(tunnelUrl)] - can be provided if isWebhookSimulated is false
+   * @param {Function} [params.removeWebhook] - can be provided if isWebhookSimulated is false
+   *
+   * Please refer to ngrok options (https://github.com/bubenshchykov/ngrok#options)
+   * @param {Object}   [params.tunnel]
+   * @param {String}   [params.tunnel.subdomain]
+   * @param {String}   [params.tunnel.auth]
+   * @param {String}   [params.tunnel.authToken]
+   */
+  constructor ({
+    t,
+    isWebhookSimulated = true,
+    simulateWebhookSince,
+    webhookFetchInterval = 2000,
+    createWebhook,
+    removeWebhook,
+    tunnel
+  }) {
+    this.t = t
+
+    // minus one second to handle cases events are generated during the same second of webhook manager creation
+    this.lastEventTimestamp = getTimestamp() - 1
+
+    this.isWebhookSimulated = isWebhookSimulated
+    this.tunnel = tunnel || {}
+
+    this.webhookFetchInterval = webhookFetchInterval
+    this.simulateWebhookSince = simulateWebhookSince
+    this.simulateWebhookInterval = null
+
+    this.createWebhook = createWebhook || noop
+    this.removeWebhook = removeWebhook || noop
+  }
+
+  async start () {
+    if (this.isWebhookSimulated) {
+      if (!this.simulateWebhookSince) {
+        throw new Error('Functions `simulateWebhookSince` expected')
+      }
+    }
+
+    if (this.isWebhookSimulated) {
+      this.simulateWebhookInterval = setInterval(async () => {
+        const timestamp = await this.simulateWebhookSince(this.lastEventTimestamp)
+        this.lastEventTimestamp = timestamp
+      }, this.webhookFetchInterval)
+    } else {
+      // https://github.com/bubenshchykov/ngrok#connect
+      this.tunnelUrl = await ngrok.connect({
+        addr: this.t.context.serverPort,
+        auth: this.tunnel.auth || undefined,
+        subdomain: this.tunnel.subdomain || undefined,
+        authtoken: this.tunnel.authToken || undefined
+      })
+
+      await this.createWebhook(this.tunnelUrl)
+    }
+  }
+
+  async stop () {
+    if (this.isWebhookSimulated) {
+      clearInterval(this.simulateWebhookInterval)
+    } else {
+      // https://github.com/bubenshchykov/ngrok#disconnect
+      await ngrok.disconnect(this.tunnelUrl)
+    }
+  }
+
+  // if isWebhookSimulated is false, really wait webhooks events
+  // otherwise, fetch events not retrieved from last time to simulate a real webhook running
+  async waitForEvents (waitDurationMs = 10000) {
```

Comment:
It seems it’s just some utility function exposed and comment above is more general advice about using `WehbookManager` than documentation about this function. In which case you should rewrite this comment and move it if needed.

Reply:
Good catch. This comment is obsolete and should be removed.